File: creator/3.1.2/engine/bin/.cache/dev/editor/transform-cache/fs/cocos/core/pipeline/shadow/shadow-flow.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ShadowFlow = void 0;

var _index = require("../../data/decorators/index.js");

var _define = require("../define.js");

var _renderFlow = require("../render-flow.js");

var _enum = require("../forward/enum.js");

var _shadowStage = require("./shadow-stage.js");

var _index2 = require("../../gfx/index.js");

var _pipelineSerialization = require("../pipeline-serialization.js");

var _shadows = require("../../renderer/scene/shadows.js");

var _sceneCulling = require("../scene-culling.js");

var _dec, _class, _class2, _temp;

/**
 * @en Shadow map render flow
 * @zh 阴影贴图绘制流程
 */
let ShadowFlow = (_dec = (0, _index.ccclass)('ShadowFlow'), _dec(_class = (_temp = _class2 = class ShadowFlow extends _renderFlow.RenderFlow {
  constructor(...args) {
    super(...args);
    this._shadowRenderPass = null;
  }

  initialize(info) {
    super.initialize(info);

    if (this._stages.length === 0) {
      // add shadowMap-stages
      const shadowMapStage = new _shadowStage.ShadowStage();
      shadowMapStage.initialize(_shadowStage.ShadowStage.initInfo);

      this._stages.push(shadowMapStage);
    }

    return true;
  }

  render(camera) {
    const pipeline = this._pipeline;
    const shadowInfo = pipeline.pipelineSceneData.shadows;
    const shadowFrameBufferMap = pipeline.pipelineSceneData.shadowFrameBufferMap;
    const shadowObjects = pipeline.pipelineSceneData.shadowObjects;

    if (!shadowInfo.enabled || shadowInfo.type !== _shadows.ShadowType.ShadowMap) {
      return;
    }

    const validLights = (0, _sceneCulling.lightCollecting)(camera, shadowInfo.maxReceived);

    if (shadowObjects.length === 0) {
      this.clearShadowMap(validLights, camera);
      return;
    }

    for (let l = 0; l < validLights.length; l++) {
      const light = validLights[l];

      if (!shadowFrameBufferMap.has(light)) {
        this._initShadowFrameBuffer(pipeline, light);
      }

      const shadowFrameBuffer = shadowFrameBufferMap.get(light);

      if (shadowInfo.shadowMapDirty) {
        this.resizeShadowMap(light, shadowInfo);
      }

      for (let i = 0; i < this._stages.length; i++) {
        const shadowStage = this._stages[i];
        shadowStage.setUsage(light, shadowFrameBuffer);
        shadowStage.render(camera);
      }
    } // After the shadowMap rendering of all lights is completed,
    // restore the ShadowUBO data of the main light.


    pipeline.pipelineUBO.updateShadowUBO(camera);
  }

  destroy() {
    super.destroy();
    const shadowFrameBufferMap = this._pipeline.pipelineSceneData.shadowFrameBufferMap;
    const shadowFrameBuffers = Array.from(shadowFrameBufferMap.values());

    for (let i = 0; i < shadowFrameBuffers.length; i++) {
      const frameBuffer = shadowFrameBuffers[i];

      if (!frameBuffer) {
        continue;
      }

      const renderTargets = frameBuffer.colorTextures;

      for (let j = 0; j < renderTargets.length; j++) {
        const renderTarget = renderTargets[i];

        if (renderTarget) {
          renderTarget.destroy();
        }
      }

      renderTargets.length = 0;
      const depth = frameBuffer.depthStencilTexture;

      if (depth) {
        depth.destroy();
      }

      frameBuffer.destroy();
    }

    shadowFrameBufferMap.clear();

    if (this._shadowRenderPass) {
      this._shadowRenderPass.destroy();
    }
  }

  _initShadowFrameBuffer(pipeline, light) {
    const device = pipeline.device;
    const shadows = pipeline.pipelineSceneData.shadows;
    const shadowMapSize = shadows.size;
    const shadowFrameBufferMap = pipeline.pipelineSceneData.shadowFrameBufferMap;
    const format = (0, _define.supportsHalfFloatTexture)(device) ? shadows.packing ? _index2.Format.RGBA8 : _index2.Format.RGBA16F : _index2.Format.RGBA8;

    if (!this._shadowRenderPass) {
      const colorAttachment = new _index2.ColorAttachment();
      colorAttachment.format = format;
      colorAttachment.loadOp = _index2.LoadOp.CLEAR; // should clear color attachment

      colorAttachment.storeOp = _index2.StoreOp.STORE;
      colorAttachment.sampleCount = 1;
      const depthStencilAttachment = new _index2.DepthStencilAttachment();
      depthStencilAttachment.format = device.depthStencilFormat;
      depthStencilAttachment.depthLoadOp = _index2.LoadOp.CLEAR;
      depthStencilAttachment.depthStoreOp = _index2.StoreOp.DISCARD;
      depthStencilAttachment.stencilLoadOp = _index2.LoadOp.CLEAR;
      depthStencilAttachment.stencilStoreOp = _index2.StoreOp.DISCARD;
      depthStencilAttachment.sampleCount = 1;
      const renderPassInfo = new _index2.RenderPassInfo([colorAttachment], depthStencilAttachment);
      this._shadowRenderPass = device.createRenderPass(renderPassInfo);
    }

    const shadowRenderTargets = [];
    shadowRenderTargets.push(device.createTexture(new _index2.TextureInfo(_index2.TextureType.TEX2D, _index2.TextureUsageBit.COLOR_ATTACHMENT | _index2.TextureUsageBit.SAMPLED, format, shadowMapSize.x, shadowMapSize.y)));
    const depth = device.createTexture(new _index2.TextureInfo(_index2.TextureType.TEX2D, _index2.TextureUsageBit.DEPTH_STENCIL_ATTACHMENT, device.depthStencilFormat, shadowMapSize.x, shadowMapSize.y));
    const shadowFrameBuffer = device.createFramebuffer(new _index2.FramebufferInfo(this._shadowRenderPass, shadowRenderTargets, depth)); // Cache frameBuffer

    shadowFrameBufferMap.set(light, shadowFrameBuffer);
  }

  clearShadowMap(validLights, camera) {
    const scene = this._pipeline.pipelineSceneData;

    for (let l = 0; l < validLights.length; l++) {
      const light = validLights[l];
      const shadowFrameBuffer = scene.shadowFrameBufferMap.get(light);

      if (!scene.shadowFrameBufferMap.has(light)) {
        continue;
      }

      for (let i = 0; i < this._stages.length; i++) {
        const shadowStage = this._stages[i];
        shadowStage.setUsage(light, shadowFrameBuffer);
        shadowStage.clearFramebuffer(camera);
      }
    }
  }

  resizeShadowMap(light, shadowInfo) {
    const width = shadowInfo.size.x;
    const height = shadowInfo.size.y;
    const pipeline = this._pipeline;
    const device = pipeline.device;
    const shadowFrameBufferMap = pipeline.pipelineSceneData.shadowFrameBufferMap;
    const format = (0, _define.supportsHalfFloatTexture)(device) ? shadowInfo.packing ? _index2.Format.RGBA8 : _index2.Format.RGBA16F : _index2.Format.RGBA8;

    if (shadowFrameBufferMap.has(light)) {
      const frameBuffer = shadowFrameBufferMap.get(light);

      if (!frameBuffer) {
        return;
      }

      const renderTargets = [];
      renderTargets.push(pipeline.device.createTexture(new _index2.TextureInfo(_index2.TextureType.TEX2D, _index2.TextureUsageBit.COLOR_ATTACHMENT | _index2.TextureUsageBit.SAMPLED, format, width, height)));
      const depth = frameBuffer.depthStencilTexture;

      if (depth) {
        depth.resize(width, height);
      }

      const shadowRenderPass = frameBuffer.renderPass;
      frameBuffer.destroy();
      frameBuffer.initialize(new _index2.FramebufferInfo(shadowRenderPass, renderTargets, depth));
    }

    shadowInfo.shadowMapDirty = false;
  }

}, _class2.initInfo = {
  name: _define.PIPELINE_FLOW_SHADOW,
  priority: _enum.ForwardFlowPriority.SHADOW,
  tag: _pipelineSerialization.RenderFlowTag.SCENE,
  stages: []
}, _temp)) || _class);
exports.ShadowFlow = ShadowFlow;

File: creator/3.1.2/engine-native/cocos/renderer/pipeline/shadow/ShadowFlow.cpp
/****************************************************************************
 Copyright (c) 2020-2021 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
 worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
 not use Cocos Creator software for developing other software or tools that's
 used for developing games. You are not granted to publish, distribute,
 sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "ShadowFlow.h"

#include "../Define.h"
#include "../forward/ForwardPipeline.h"
#include "../helper/SharedMemory.h"
#include "ShadowStage.h"
#include "gfx-base/GFXDevice.h"
#include "gfx-base/GFXFramebuffer.h"
#include "gfx-base/GFXRenderPass.h"
#include "gfx-base/GFXTexture.h"
#include "../SceneCulling.h"

namespace cc::pipeline {
RenderFlowInfo ShadowFlow::initInfo = {
    "ShadowFlow",
    static_cast<uint>(ForwardFlowPriority::SHADOW),
    static_cast<uint>(RenderFlowTag::SCENE),
    {},
};
const RenderFlowInfo &ShadowFlow::getInitializeInfo() { return ShadowFlow::initInfo; }

ShadowFlow::~ShadowFlow() = default;

bool ShadowFlow::initialize(const RenderFlowInfo &info) {
    RenderFlow::initialize(info);
    if (_stages.empty()) {
        auto *shadowStage = CC_NEW(ShadowStage);
        shadowStage->initialize(ShadowStage::getInitializeInfo());
        _stages.emplace_back(shadowStage);
    }

    return true;
}

void ShadowFlow::activate(RenderPipeline *pipeline) {
    RenderFlow::activate(pipeline);
}

void ShadowFlow::render(Camera *camera) {
    const auto *sceneData = _pipeline->getPipelineSceneData();
    const auto *shadowInfo = sceneData->getSharedData()->getShadows();
    if (!shadowInfo->enabled || shadowInfo->getShadowType() != ShadowType::SHADOWMAP) return;

    lightCollecting(camera, _validLights);

    if (sceneData->getShadowObjects().empty()) {
        clearShadowMap(camera);
        return;
    }

    const auto &shadowFramebufferMap = sceneData->getShadowFramebufferMap();
    for (const auto *light : _validLights) {
        if (!shadowFramebufferMap.count(light)) {
            initShadowFrameBuffer(_pipeline, light);
        }

        auto *shadowFrameBuffer = shadowFramebufferMap.at(light);
        if (shadowInfo->shadowMapDirty) {
            resizeShadowMap(light, shadowInfo);
        }
        for (auto *stage : _stages) {
            auto *shadowStage = dynamic_cast<ShadowStage *>(stage);
            shadowStage->setUseData(light, shadowFrameBuffer);
            shadowStage->render(camera);
        }
    }

    // After the shadowMap rendering of all lights is completed,
    // restore the ShadowUBO data of the main light.
    _pipeline->getPipelineUBO()->updateShadowUBO(camera);
}

void ShadowFlow::clearShadowMap(Camera *camera) {
    auto *sceneData = _pipeline->getPipelineSceneData();
    const auto &shadowFramebufferMap = sceneData->getShadowFramebufferMap();
    for (const auto *light : _validLights) {
        if (!shadowFramebufferMap.count(light)) {
            continue;
        }

        auto *shadowFrameBuffer = shadowFramebufferMap.at(light);
        for (auto *stage : _stages) {
            auto *shadowStage = dynamic_cast<ShadowStage *>(stage);
            shadowStage->setUseData(light, shadowFrameBuffer);
            shadowStage->clearFramebuffer(camera);
        }
    }
}

void ShadowFlow::resizeShadowMap(const Light *light, const Shadows *shadowInfo){
    auto *sceneData = _pipeline->getPipelineSceneData();
    auto *     device    = gfx::Device::getInstance();
    const auto width     = static_cast<uint>(shadowInfo->size.x);
    const auto height    = static_cast<uint>(shadowInfo->size.y);
    const auto format    = supportsHalfFloatTexture(device)
                            ? (shadowInfo->packing ? gfx::Format::RGBA8 : gfx::Format::RGBA16F)
                            : gfx::Format::RGBA8;

    if (sceneData->getShadowFramebufferMap().count(light)) {
        auto *framebuffer = sceneData->getShadowFramebufferMap().at(light);

        if (!framebuffer) {
            return;
        }

        auto renderTargets = framebuffer->getColorTextures();
        for (auto *renderTarget : renderTargets) {
            CC_DELETE(renderTarget);
        }
        renderTargets.clear();
        renderTargets.emplace_back(gfx::Device::getInstance()->createTexture({
            gfx::TextureType::TEX2D,
            gfx::TextureUsageBit::COLOR_ATTACHMENT | gfx::TextureUsageBit::SAMPLED,
            format,
            width,
            height,
        }));
        for (auto *renderTarget : renderTargets) {
            _usedTextures.emplace_back(renderTarget);
        }

        auto *depth = framebuffer->getDepthStencilTexture();
        CC_DELETE(depth);
        depth = device->createTexture({
            gfx::TextureType::TEX2D,
            gfx::TextureUsageBit::DEPTH_STENCIL_ATTACHMENT,
            device->getDepthStencilFormat(),
            width,
            height,
        });
        _usedTextures.emplace_back(depth);

        framebuffer->destroy();
        framebuffer->initialize({
            _renderPass,
            renderTargets,
            depth,
            {},
        });
    }
}

void ShadowFlow::initShadowFrameBuffer(RenderPipeline *pipeline, const Light *light) {
    auto *      device        = gfx::Device::getInstance();
    const auto *sceneData     = _pipeline->getPipelineSceneData();
    const auto *shadowInfo    = sceneData->getSharedData()->getShadows();
    const auto  shadowMapSize = shadowInfo->size;
    const auto  width         = static_cast<uint>(shadowMapSize.x);
    const auto  height        = static_cast<uint>(shadowMapSize.y);
    const auto  format        = supportsHalfFloatTexture(device)
                            ? (shadowInfo->packing ? gfx::Format::RGBA8 : gfx::Format::RGBA16F)
                            : gfx::Format::RGBA8;

    if (!_renderPass) {
        const gfx::ColorAttachment colorAttachment = {
            format,
            gfx::SampleCount::X1,
            gfx::LoadOp::CLEAR,
            gfx::StoreOp::STORE,
            {},
            {},
        };

        const gfx::DepthStencilAttachment depthStencilAttachment = {
            device->getDepthStencilFormat(),
            gfx::SampleCount::X1,
            gfx::LoadOp::CLEAR,
            gfx::StoreOp::DISCARD,
            gfx::LoadOp::CLEAR,
            gfx::StoreOp::DISCARD,
            {},
            {},
        };

        gfx::RenderPassInfo rpInfo;
        rpInfo.colorAttachments.emplace_back(colorAttachment);
        rpInfo.depthStencilAttachment = depthStencilAttachment;
        _renderPass                   = device->createRenderPass(rpInfo);
    }

    vector<gfx::Texture *> renderTargets;
    renderTargets.emplace_back(device->createTexture({
        gfx::TextureType::TEX2D,
        gfx::TextureUsageBit::COLOR_ATTACHMENT | gfx::TextureUsageBit::SAMPLED,
        format,
        width,
        height,
    }));
    for (auto *renderTarget : renderTargets) {
        _usedTextures.emplace_back(renderTarget);
    }

    gfx::Texture *depth = device->createTexture({
        gfx::TextureType::TEX2D,
        gfx::TextureUsageBit::DEPTH_STENCIL_ATTACHMENT,
        device->getDepthStencilFormat(),
        width,
        height,
    });
    _usedTextures.emplace_back(depth);

    gfx::Framebuffer *framebuffer = device->createFramebuffer({
        _renderPass,
        renderTargets,
        depth,
        {}, //colorMipmapLevels
    });

    pipeline->getPipelineSceneData()->setShadowFramebuffer(light, framebuffer);
}

void ShadowFlow::destroy() {
    CC_SAFE_DESTROY(_renderPass)

    for (auto *texture : _usedTextures) {
        CC_SAFE_DESTROY(texture);
    }
    _usedTextures.clear();

    _validLights.clear();

    RenderFlow::destroy();
}

} // namespace cc::pipeline
